src/Soil-Core/SoilBehaviorRegistry.class.st

```diff
@@ -16,24 +16,29 @@ SoilBehaviorRegistry >> addSpecialObjects [
 		put: SOBehaviorDescription metaId
 ]
 
+{ #category : #'as yet unclassified' }
+SoilBehaviorRegistry >> behaviorDescriptionWithIndex: behaviorIndex andVersion: version transaction: transaction [
+	| versionsOfBehavior |
+
+	self loadHistoryForBehaviorWithId: (SOObjectId segment: 0 index: behaviorIndex) transaction: transaction.
+
+	versionsOfBehavior := versions at: behaviorIndex ifAbsent: [self halt].
+
+	^versionsOfBehavior
+		detect: [ :behav | behav version >= version ]
+		ifNone: [(SOObjectNotFound new segment: 0; index: behaviorIndex) signal]
+]
+
 { #category : #'as yet unclassified' }
 SoilBehaviorRegistry >> behaviorVersionsUpTo: aSOBehaviorDescription transaction: transaction [
-	| objectId records chain offset |
+	| objectId chain offset |
 	objectId := self 
 		nameAt: aSOBehaviorDescription behaviorIdentifier 
 		ifAbsent: [ self halt ].
-	records := soil objectRepository 
-		allVersionsOf: objectId.
-	chain := records collect: [ :record |
-		record 
-			transaction: transaction;
-			materializeObject;
-			object  ].
+	self loadHistoryForBehaviorWithId: objectId transaction: transaction.
+	chain := versions at: objectId index.
 	chain first isCurrent ifFalse: [ 
-		chain addFirst: (SOBehaviorDescription for: aSOBehaviorDescription objectClass) ].
-	versions 
-		at: aSOBehaviorDescription behaviorIdentifier 
-		put: chain.
+		chain addFirst: ((SOBehaviorDescription for: aSOBehaviorDescription objectClass) version: (aSOBehaviorDescription version + 1))].
```

Comment:
What happens in case the code is newer than the one in the database? On read time we add the current version on top but on writing an object with the new version who is storing the behavior description. Can it be that we end up with an object in database that has a behavior description version missing?

Reply:
this is only done if the code is newer (if it changed):

```
	chain first isCurrent ifFalse: [ 
		chain addFirst: ((SOBehaviorDescription for: aSOBehaviorDescription objectClass) version: (aSOBehaviorDescription version + 1))].
```

Reply:
That is what I meant. I you first read an object with a newer code the new behavior description is added to the top of the chain. But if after that an object is written the behavior description does not get written to database, right?

Reply:
I improved SoilMigrationTest>>#testMaterializingObjectWithIvarRemovedThenCommit, we store an object, then remove an ivar, then load it back, then we commit it again and it works:

```
testMaterializingObjectWithIvarRemovedThenCommit
	"We can load an object that was saved with two ivars even if the current class has just one"
	| tx tx2 materializedRoot object |
	object := migrationClass new.
	object
		instVarNamed: #one put: 1;
		instVarNamed: #two put: 2.
	tx := soil newTransaction.
	tx root: object.
	tx commit.
	migrationClass
		removeSlot: (migrationClass slotNamed: #two).
	
	"load it"	
	tx2 := soil newTransaction.
	materializedRoot := tx2 root.
	self assert: materializedRoot class instVarNames size equals: 1.
	self assert: (materializedRoot instVarNamed: #one) equals: 1.

	"lets try to commit with the class changed"
	tx := soil newTransaction.
	tx root: materializedRoot.
	tx commit.

	"load again"
	tx2 := soil newTransaction.
	materializedRoot := tx2 root.
	self assert: materializedRoot class instVarNames size equals: 1.
	self assert: (materializedRoot instVarNamed: #one) equals: 1
```

Reply:
Sure it works because the change class is still the same as when writing it. So a missing behavior description in the database cannot be seen because the one created from memory will be the same. The test should test if there are two versions of that behavior are in the database